contracts/libraries/ArbitraryMessage.sol

```diff
@@ -4,56 +4,63 @@ import "../interfaces/IBridgeValidators.sol";
-    // when reading from offset 20 mload will read 12 zero bytes followed
-    // by the 20 recipient address bytes and correctly convert it into an address.
-    // this saves some storage/gas over the alternative solution
-    // which is padding address to 32 bytes and reading recipient at offset 32.
-    // for more details see discussion in:
-    // https://github.com/paritytech/parity-bridge/issues/61
+    /**
+    * @dev Unpacks data fields from AMB message
+    * layout of message :: bytes:
+    * offset  0: 32 bytes :: uint256 - message length
+    * offset 32: 32 bytes :: bytes32 messageId
+    * offset 64: 32 bytes :: uint256 chainId
+    * offset 96: 20 bytes :: address - sender address
+    * offset 116: 20 bytes :: address - executor contract
+    * offset 136: 4 bytes :: uint32 - gasLimit
+    * offset 140: 1 bytes :: bytes1 - dataType
+    * (optional) 141: 32 bytes :: uint256 - gasPrice
+    * (optional) 141: 1 bytes :: bytes1 - gasPriceSpeed
 
+    * bytes 1 to 32 are 0 because message length is stored as little endian.
+    * mload always reads 32 bytes.
+    * so we can and have to start reading recipient at offset 20 instead of 32.
+    * if we were to read at 32 the address would contain part of value and be corrupted.
+    * when reading from offset 20 mload will read 12 zero bytes followed
+    * by the 20 recipient address bytes and correctly convert it into an address.
+    * this saves some storage/gas over the alternative solution
+    * which is padding address to 32 bytes and reading recipient at offset 32.
+    * for more details see discussion in:
+    * https://github.com/paritytech/parity-bridge/issues/61
+    * @param _data encoded message
+    * @param applyDataOffset indicates if a calldata contains a second argument after message itself
+    */
     function unpackData(bytes _data, bool applyDataOffset)
         internal
         pure
         returns (
+            bytes32 messageId,
+            uint256 chainId,
             address sender,
             address executor,
-            bytes32 txHash,
-            uint256 gasLimit,
+            uint32 gasLimit,
             bytes1 dataType,
             uint256 gasPrice,
             bytes memory data
         )
     {
         uint256 dataOffset = 0;
         uint256 datasize;
-        // 32 (tx hash) + 20 (sender)  + 20 (executor) + 32 (gasLimit) + 1 (dataType)
-        uint256 srcdataptr = 32 + 20 + 20 + 32 + 1;
+        // 32 (message id) + 32 (chain id) + 20 (sender)  + 20 (executor) + 4 (gasLimit) + 1 (dataType)
+        uint256 srcdataptr = 32 + 32 + 20 + 20 + 4 + 1;
         assembly {
-            txHash := mload(add(_data, 32))
-            sender := mload(add(_data, 52))
-            executor := mload(add(_data, 72))
-            gasLimit := mload(add(_data, 104))
-            dataType := and(mload(add(_data, 136)), 0xFF00000000000000000000000000000000000000000000000000000000000000)
+            messageId := mload(add(_data, 32))
```

Comment:
Does it make sense to check that first 4 bytes of the message id match with `ENCODED_BRIDGE_VERSION`? If not, it seems that it is not the message we need to handle.

Reply:
Added such checks in `BasicHomeAMB.sol` and `ForeignHomeAMB.sol` in e35136b

Reply:
Do you think it makes sense to introduce a separate method for this to simplify the code?